internal/sourcemap/sourcemap_fetcher.go

```diff
@@ -0,0 +1,125 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package sourcemap
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+
+	"github.com/go-sourcemap/sourcemap"
+
+	"github.com/elastic/apm-server/internal/logs"
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+type SourcemapFetcher struct {
+	metadata MetadataFetcher
+	backend  Fetcher
+	logger   *logp.Logger
+}
+
+func NewSourcemapFetcher(metadata MetadataFetcher, backend Fetcher) *SourcemapFetcher {
+	s := &SourcemapFetcher{
+		metadata: metadata,
+		backend:  backend,
+		logger:   logp.NewLogger(logs.Sourcemap),
+	}
+
+	return s
+}
+
+func (s *SourcemapFetcher) Fetch(ctx context.Context, name, version, path string) (*sourcemap.Consumer, error) {
+	original := Identifier{name: name, version: version, path: path}
+
+	select {
+	case <-s.metadata.Ready():
+		// the mutex is shared by the update goroutine, we need to release it
+		// as soon as possible to avoid blocking updates.
+		if i, ok := s.metadata.GetID(original); ok {
+			// Only fetch from ES if the sourcemap id exists
+			return s.fetch(ctx, i)
+		}
+	default:
+		s.logger.Debugf("Metadata cache not populated. Falling back to backend fetcher for id: %s, %s, %s", name, version, path)
+		// init is in progress, ignore the metadata cache and fetch the sourcemap directly
+		// return if we get a valid sourcemap or an error
+		if c, err := s.backend.Fetch(ctx, original.name, original.version, original.path); c != nil || err != nil {
+			return c, err
+		}
```

Comment:
I'm not 100% convinced about this fallback. Running your code, if an unreachable Elasticsearch is configured, this is what is logged:
```
{"log.level":"debug","@timestamp":"2023-02-02T08:35:02.675+0100","log.logger":"sourcemap","log.origin":{"file.name":"sourcemap/sourcemap_fetcher.go","file.line":59},"message":"Metadata cache not populated. Falling back to backend fetcher for id: apm-agent-js, 1.0.1, http://localhost:8000/test/e2e/general-usecase/bundle.js.map","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"error","@timestamp":"2023-02-02T08:35:03.229+0100","log.logger":"beater","log.origin":{"file.name":"beater/waitready.go","file.line":64},"message":"precondition failed: dial tcp [::1]:9200: connect: connection refused","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-02-02T08:35:07.676+0100","log.logger":"stacktrace","log.origin":{"file.name":"sourcemap/processor.go","file.line":124},"message":"failed to fetch sourcemap with path (http://localhost:8000/test/e2e/general-usecase/bundle.js.map): failure querying ES: context deadline exceeded","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-02-02T08:35:07.676+0100","log.logger":"sourcemap","log.origin":{"file.name":"sourcemap/sourcemap_fetcher.go","file.line":59},"message":"Metadata cache not populated. Falling back to backend fetcher for id: apm-agent-js, 1.0.1, http://localhost:8000/test/e2e/general-usecase/bundle.js.map","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-02-02T08:35:07.676+0100","log.logger":"stacktrace","log.origin":{"file.name":"sourcemap/processor.go","file.line":124},"message":"failed to fetch sourcemap with path (http://localhost:8000/test/e2e/general-usecase/bundle.js.map): failure querying ES: context deadline exceeded","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-02-02T08:35:07.676+0100","log.logger":"sourcemap","log.origin":{"file.name":"sourcemap/sourcemap_fetcher.go","file.line":59},"message":"Metadata cache not populated. Falling back to backend fetcher for id: apm-agent-js, 1.0.1, http://localhost:8000/test/e2e/general-usecase/bundle.js.map","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-02-02T08:35:07.676+0100","log.logger":"stacktrace","log.origin":{"file.name":"sourcemap/processor.go","file.line":124},"message":"failed to fetch sourcemap with path (http://localhost:8000/test/e2e/general-usecase/bundle.js.map): failure querying ES: context deadline exceeded","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-02-02T08:35:07.676+0100","log.logger":"sourcemap","log.origin":{"file.name":"sourcemap/sourcemap_fetcher.go","file.line":59},"message":"Metadata cache not populated. Falling back to backend fetcher for id: apm-agent-js, 1.0.1, http://localhost:8000/test/e2e/general-usecase/bundle.js.map","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-02-02T08:35:07.676+0100","log.logger":"stacktrace","log.origin":{"file.name":"sourcemap/processor.go","file.line":124},"message":"failed to fetch sourcemap with path (http://localhost:8000/test/e2e/general-usecase/bundle.js.map): failure querying ES: context deadline exceeded","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-02-02T08:35:07.676+0100","log.logger":"sourcemap","log.origin":{"file.name":"sourcemap/sourcemap_fetcher.go","file.line":59},"message":"Metadata cache not populated. Falling back to backend fetcher for id: apm-agent-js, 1.0.1, http://localhost:8000/test/e2e/general-usecase/bundle.js.map","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-02-02T08:35:07.676+0100","log.logger":"stacktrace","log.origin":{"file.name":"sourcemap/processor.go","file.line":124},"message":"failed to fetch sourcemap with path (http://localhost:8000/test/e2e/general-usecase/bundle.js.map): failure querying ES: context deadline exceeded","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-02-02T08:35:07.676+0100","log.logger":"sourcemap","log.origin":{"file.name":"sourcemap/sourcemap_fetcher.go","file.line":59},"message":"Metadata cache not populated. Falling back to backend fetcher for id: apm-agent-js, 1.0.1, http://localhost:8000/test/e2e/general-usecase/bundle.js.map","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-02-02T08:35:07.676+0100","log.logger":"stacktrace","log.origin":{"file.name":"sourcemap/processor.go","file.line":124},"message":"failed to fetch sourcemap with path (http://localhost:8000/test/e2e/general-usecase/bundle.js.map): failure querying ES: context deadline exceeded","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-02-02T08:35:07.678+0100","log.logger":"request","log.origin":{"file.name":"middleware/log_middleware.go","file.line":61},"message":"request accepted","service.name":"apm-server","url.original":"/intake/v2/rum/events","http.request.method":"POST","user_agent.original":"curl/7.79.1","source.address":"127.0.0.1","http.request.id":"9dab8633-386a-47c6-b431-171ba122b9b9","event.duration":5003973201,"http.response.status_code":202,"ecs.version":"1.6.0"}
{"log.level":"error","@timestamp":"2023-02-02T08:35:13.223+0100","log.logger":"agentcfg","log.origin":{"file.name":"agentcfg/elasticsearch.go","file.line":134},"message":"refresh cache error: context deadline exceeded","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"error","@timestamp":"2023-02-02T08:35:18.234+0100","log.logger":"beater","log.origin":{"file.name":"beater/waitready.go","file.line":64},"message":"precondition failed: dial tcp [::1]:9200: connect: connection refused","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"error","@timestamp":"2023-02-02T08:35:23.221+0100","log.logger":"sourcemap","log.origin":{"file.name":"sourcemap/metadata_fetcher.go","file.line":117},"message":"failed to sync sourcemaps metadata: failed to run initial search query: context deadline exceeded","service.name":"apm-server","ecs.version":"1.6.0"}
```

The fallback only makes sense if ES is reachable & accessible, but the background sync job is not yet finished.

Reply:
Good catch! I've added a test for ES unreachable.

The current issue was the ES fetcher not returning errFetcherUnavailable when ES is not available.

The problems is that by the time the sourcemap fetcher realizes ES is not reachable/available the context is expired as you can see from your log message. That means the next fetcher will fail regardless since the deadline is exceeded.
Should we create a new context with a 1s (?) timeout when trying the next chained fetcher ?

Reply:
I started the APM Server with pointing to a non-existent ES instance, and still see the above logs. The same behavior happens when configuring another output than ES. In such cases, or if the configured credentials do not have the right privileges, the Kibana fallback fetcher should be used. The extra query to ES for every error stacktrace frame before using the Kibana fetcher is not efficient.

Reply:
That's strange, I'll investigate a bit more!

The wrong credential and ES unreachable should be covered by the systemtests: https://github.com/elastic/apm-server/pull/9722/files#diff-d98c3b8e30f1c6e2ae6dfb07a2b2f35fd940f74fd52149b5b25bf648aa622b9cR182-R237

Reply:
Alternatively, I think it would be acceptable if the server doesn't apply source maps during startup while the cache is empty. That would be simplest.

Reply:
+1 - I also think it is acceptable, as the initial sourcemap metadata fetching should be triggered immediately and be fairly quick.

Reply:
Updated to block until the metadata cache is populated